File: searching-algorithms/02-binary-search/binary-search.js

const binarySearch = (arr, num) => {
  // Divide and conquer

  let left = 0;
  let right = arr.length - 1;

  while (left <= right) {
    const mid = Math.floor((left + right) / 2);
    if (num === arr[mid]) return mid;
    else if (num < arr[mid]) right = mid - 1;
    else if (num > arr[mid]) left = mid + 1;
  }
  return -1;
};

console.log(binarySearch([1, 2, 4, 6, 8, 9, 11, 14, 16, 19, 22, 57], 19));  // 9
console.log(binarySearch([1, 4, 5, 6, 8, 9, 10, 12], 3));  // -1

File: searching-algorithms/02-binary-search/binary-search.py
#!/usr/bin/env python3


def binary_search(lst, num):

    left = 0
    right = len(lst) - 1

    while left <= right:
        mid = (left + right) // 2
        if num == lst[mid]:
            return mid
        if num < lst[mid]:
            right = mid - 1
        elif num > lst[mid]:
            left = mid + 1
    return -1


print(binary_search([1, 2, 4, 6, 8, 9, 11, 14, 16, 19, 22, 57], 19))  # 9
print(binary_search([1, 4, 5, 6, 8, 9, 10, 12], 3))  # -1
